src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioProjectEngineFactoryProvider.cs

```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.ProjectEngineHost;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor;
+
+[Export(typeof(IProjectEngineFactoryProvider))]
```

Comment:
Does this not need some kind of `Import` or `ImportingConstructor` attribute for the export to work? Or is this a v1/v2 difference thing I am about to learn (like the Shared thing I learned last week)

Reply:
No. FWIW, I don't believe any version of MEF requires something to be imported on an export. Actually thinking about it for a moment, I don't think it's possible for MEF to require every Export to have an Import without creating circularities.

Reply:
Maybe this is me unknowingly being part of a cargo cult, but I could have sworn I've needed to add (empty) importing constructors to get the composition to work at various points. Will keep an eye out for it next time :)

Reply:
Thankfully, MEF doesn't work that way. `[Export]` causes a type to be instantiated and included as part of the composition. That exported thing can import other parts of the composition, but it is not required to do so.

Here's a breakpoint being hit to show `VisualStudioProjectEngineFactoryProvider` is included in the MEF composition and can be imported elsewhere, even though it doesn't import anything and just has a public, parameterless constructor:

![image](https://github.com/dotnet/razor/assets/116161/d4c2f3b6-0621-4e11-b1fa-6c6a0a90af36)

----

Since `VisualStudioProjectEngineFactoryProvider` is defined with a MEFv1 `[Export]`, here's an example to show that a simple `[Export]` with no imports also works as expected in MEFv2:

![image](https://github.com/dotnet/razor/assets/116161/5ca2034a-b9f3-49dd-b1d0-059745379c9b)

----

There are a number of differences between MEFv1 and MEFv2, but the one that most folks stub their toe on is that MEFv1 creates shared parts by default and MEFv2 does not.

Reply:
Here's another test to show why that `[Shared]` attribute is so critical in MEFv2:

![image](https://github.com/dotnet/razor/assets/116161/70bb6c4a-0914-432e-a7f1-03b1f4a9132e)

Reply:
Just to close the loop, here's a test demonstrating the reverse "shared" behavior in MEFv1. In MEFv1, you have to actually state that an export is "non-shared":

![image](https://github.com/dotnet/razor/assets/116161/fccfe018-eb54-4629-80cd-04848d9b9035)